ktor-client/ktor-client-darwin/darwin/src/io/ktor/client/engine/darwin/DarwinWebsocketDelegate.kt

```diff
@@ -0,0 +1,202 @@
+
+    override val outgoing: SendChannel<Frame>
+        get() = _outgoing
+
+    override val extensions: List<WebSocketExtension<*>>
+        get() = emptyList()
+
+    public val delegate: NSURLSessionWebSocketDelegateProtocol = WebSocketDelegate()
+
+    init {
+        launch {
+            receiveMessages()
+        }
+        launch {
+            sendMessages()
+        }
+        GlobalScope.launch(coroutineContext) {
+            try {
+                this.coroutineContext[Job]!!.join()
+            } catch (cause: Exception) {
+                val code = CloseReason.Codes.INTERNAL_ERROR.code.convert<NSInteger>()
+                task.cancelWithCloseCode(code, "Client failed".toByteArray().toNSData())
+            } finally {
+                _incoming.close()
+                _outgoing.cancel()
+            }
+        }
+    }
+
+    private suspend fun receiveMessages() {
+        val message = withContext(receiveJob) {
+            suspendCancellableCoroutine<NSURLSessionWebSocketMessage?> {
+                task.receiveMessageWithCompletionHandler { message, error ->
+                    if (error != null) {
+                        it.resumeWithException(DarwinHttpRequestException(error))
+                        return@receiveMessageWithCompletionHandler
+                    }
+
+                    it.resume(message)
+                }
+            }
+        }
+        val frame = when (message?.type) {
+            NSURLSessionWebSocketMessageTypeData ->
+                Frame.Binary(true, message.data()!!.toByteArray())
+            NSURLSessionWebSocketMessageTypeString ->
+                Frame.Text(true, message.string()!!.toByteArray())
+            else -> throw IllegalArgumentException("Unknown message $message")
+        }
+        _incoming.send(frame)
+        receiveMessages()
+    }
+
+    private suspend fun sendMessages() {
+        _outgoing.consumeEach { frame ->
+            when (frame.frameType) {
+                FrameType.TEXT -> {
+                    suspendCancellableCoroutine<Unit> { continuation ->
+                        task.sendMessage(NSURLSessionWebSocketMessage(String(frame.data))) { error ->
+                            if (error == null) continuation.resume(Unit)
+                            else continuation.resumeWithException(DarwinHttpRequestException(error))
+                        }
+                    }
+                }
+                FrameType.BINARY -> {
+                    suspendCancellableCoroutine<Unit> { continuation ->
+                        task.sendMessage(NSURLSessionWebSocketMessage(frame.data.toNSData())) { error ->
+                            if (error == null) continuation.resume(Unit)
+                            else continuation.resumeWithException(DarwinHttpRequestException(error))
+                        }
+                    }
+                }
+                FrameType.CLOSE -> {
+                    val data = buildPacket { writeFully(frame.data) }
```

Comment:
why is it needed to first create ByteReadPacket from ByteArray, and 2 lines after convert it back to ByteArray?

Reply:
To read the first two bytes of it, which represents close reason. I can do it manually, but it won't make much difference here.